userspace/libsinsp/container_engine/cri.cpp

```diff
@@ -242,7 +243,27 @@ bool cri::resolve(sinsp_threadinfo *tinfo, bool query_os_for_missing_info)
 			g_logger.format(sinsp_logger::SEV_DEBUG,
 					"cri_async (%s): Starting synchronous lookup",
 					container_id.c_str());
+			// lookup_sync function directly invokes the container engine specific parser `parse`
 			done = m_async_source->lookup_sync(key, result);
+			// explicitly check for the most crucial retrieved value to be present
+			if(!result.m_image.empty())
+			{
+				/*
+				* Only for synchronous lookup option (e.g. Falco's default is async not sync)
+				*
+				* Fast-track addition of enriched containers (fields successfully retrieved from the container runtime socket)
+				* to the container cache, bypassing the round-trip process:
+				* `source_callback` -> `notify_new_container` ->
+				* `container_to_sinsp_event(container_to_json(container_info), ...)` ->
+				* `parse_container_json_evt` -> `m_inspector->m_container_manager.add_container()`
+				*
+				* Although we still re-add the container in `parse_container_json_evt` to also support native 'container' events, it
+				* introduces an avoidable delay in the incoming syscall event stream. Syscall events do not explicitly require container
+				* events and instead directly retrieve container details from the container cache. This behavior could potentially
+				* contribute to the issues noted by adopters, such as the absence of container images in syscall events.
+				*/
+				cache->add_container(std::make_shared<sinsp_container_info>(result), nullptr);
```

Comment:
Maybe we should introduce a new function as this one also loops over callbacks. In general, I dislike how certain functions are overloaded in the container engine. 

CC @gnosek ?

Reply:
Using existing `replace_container` instead. This should be safe and do what we want.
@mstemm would you have additional info on the callback or the general flow here? Eager to understand subtleties better. Thank you!

Reply:
I'm not sure why we even have a synchronous lookup option. That would surely lead to event drops wouldn't it? That's why we made the lookups asynchronous in the first place.

If you're asking how the asynchronous lookup flow works, you got the general steps right:
1. The container engine does an asynchronous lookup of container metadata.
2. The container metadata is packed into a CONTAINER_JSON event.
3. The event is enqueued to the inspector by the container engine thread.
4. On the next call to sinsp::next() by the event processing thread, the enqueued event is read and parsed. Parsing the event adds the container info to libsinsp.

I think step 2 is probably more confusing than it needs to be.

Reply:
The `async` is not working well in production, you have many events with missing container enrichment. My follow up question is that synchronous still isn't truly synchronous? We do the CRI API calls and parsing synchronously but then we don't add the container right away to the cache and still do the container event round trip before we add the container to the container cache. WDYT?

Reply:
I thought all the container engine code lookups were done in a separate thread, using the mechanisms in https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/container_engine/container_async_source.tpp. The reason for creating the CONTAINER_JSON event and enqueuing it to the inspector is because it's a separate thread and we can't directly access the libsinsp state, due to no locking/mutexes at all for those data structures.

There may be bugs in the async lookup path, and I agree that the current code is unnecessarily confusing, but you definitely don't want to go back to synchronous lookups as a solution for that. Some container engines had a delay of up to 500ms before container info were available for a new container, and stalling the event loop was definitely causing dropped syscalls while the synchronous lookup was underway.

Reply:
> we can't directly access the libsinsp state

We can definitely access the container cache that is protected by a Mutex Guard. We don't need more than that for the most crucial container infos. By the way I am unclear about the state wrt to container users. users could be tackled later.

---

I see 2 follow up discussions:
1. When to store the container after parsing (currently it seems not optimal to store it after doing all this round-trip stuff (JSON back and forth and after parsing the enqueued container event) which in a way seems added very unnecessary delays when we could have done it already right away,)
2. General async vs sync lookups. 

Urgency: Complains about container engine enrichment issues date back over 3-5 years, I would say it's time to improve the container engine.

Reply:
It's great that the container info has protection now. It's still advantageous to create a CONTAINER_JSON event, though, as that can be written to trace files. That allows reading the trace file later and getting the container info. Without that event, an application reading a trace file will never have container info for containers that were created while the capture was running.

I wouldn't actually think there is much delay on top of the asynchronous lookup itself, btw. Serializing the container info into an object is very fast, and as soon as it's queued to the inspector, it will be read and parsed as the very next event.

Definitely support a reworking of all this code to make it easier to understand.

Reply:
oh oh another instance of trace files causing interference w/ real-time processing. Are there ways to separate these use cases? Sometimes I feel scap_files are in the way of progress / innovation and real-time stream performance for actual use cases in prod ....

> as soon as it's queued to the inspector, it will be read and parsed as the very next event.

just on my localhost with one pod I could simulate improvements, so I think we should quantify these claims with data. I believe it does add avoidable delays. Also just from a logic perspective it seems we can come up with a better design and it sounds you would be on board to rework things.

What would be best next steps from your perspective?

Reply:
I guess the first step would be identifying a champion that can design an alternative design and present it. (Unfortunately, I don't think I can be that champion, but I'd be more than willing to provide advice and reviews.)

Reply:
If no one else can or wants to sure I can help. In this case since we talk about the container engine "master of disaster" seems more accurate. 

I think as a first step we should start with storing the container before we convert to JSON in any case. That way we don't break the trace files and delays don't matter for the native container events anyways, but storing the container as fast as possible does matter a lot for the syscalls event stream.

After that we can rework the more general flow. 

Sounds reasonable?

Reply:
Yeah, if the container cache is protected across threads it makes sense to add it immediately and then create the CONTAINER_JSON event afterward.

Reply:
ok one more question: what do the callbacks do or can do? Was there a reason the callback should only be called from the enqueued container event?

Reply:
I guess you mean the on_new_container callback? I believe it was a general mechanism for other parts of libsinsp to be notified when there is a new container. And for this, it's used to pass back control to from the container engines to libsinsp (I think, I haven't looked at the code in a while).

Reply:
ok, sounds like no particular reason why they were invoked from the enqueued event right? Hence we could upfront them as well, I think other container engines do that. I will look into it.